internal/testrunner/runners/pipeline/runner.go

```diff
@@ -103,9 +111,27 @@ func (r *runner) run() ([]testrunner.TestResult, error) {
 		return nil, fmt.Errorf("failed to read data stream manifest: %w", err)
 	}
 
-	expectedDataset := dsManifest.Dataset
-	if expectedDataset == "" {
-		expectedDataset = pkgManifest.Name + "." + r.options.TestFolder.DataStream
+	// when reroute processors are used, expectedDatasets should be set depends on the processor config
+	var expectedDatasets []string
+	for _, pipeline := range r.pipelines {
+		var esIngestPipeline IngestPipelineReroute
+		err = yaml.Unmarshal(pipeline.Content, &esIngestPipeline)
+		if err != nil {
+			return nil, fmt.Errorf("unmarshalling ingest pipeline content failed: %w", err)
+		}
+		for _, processor := range esIngestPipeline.Processors {
+			if reroute, ok := processor["reroute"]; ok {
+				expectedDatasets = append(expectedDatasets, reroute.Dataset...)
+			}
+		}
+	}
+
+	if expectedDatasets == nil {
```

Comment:
Nit. Just in case it is initialized in some refactor.
```suggestion
	if len(expectedDatasets) == 0 {
```